pkgs/servers/http/h2o/nix-store-etag.patch

```diff
@@ -0,0 +1,137 @@
+From f08b066f2bfb6f93e8657366d65f29aa7d7cba53 Mon Sep 17 00:00:00 2001
+From: Emily <vcs@emily.moe>
+Date: Tue, 31 Mar 2020 19:22:06 +0100
+Subject: [PATCH] use hash of path as etag in Nix store
+
+---
+ include/h2o/filecache.h |  6 ++++-
+ lib/common/filecache.c  | 54 ++++++++++++++++++++++++++++++++++++++++-
+ lib/handler/file.c      |  8 +++---
+ 3 files changed, 63 insertions(+), 5 deletions(-)
+
+diff --git a/include/h2o/filecache.h b/include/h2o/filecache.h
+index 064e20e9..a845c83a 100644
+--- a/include/h2o/filecache.h
++++ b/include/h2o/filecache.h
+@@ -29,7 +29,10 @@
+ #include "h2o/memory.h"
+ #include "h2o/time_.h"
+ 
+-#define H2O_FILECACHE_ETAG_MAXLEN (sizeof("\"deadbeef-deadbeefdeadbeef\"") - 1)
++/*   28 bytes of truncated hash
++ * ×  2 for hex encoding
++ * +  2 bytes of double quotes */
++#define H2O_FILECACHE_ETAG_MAXLEN ((28 * 2) + 2)
+ 
+ typedef struct st_h2o_filecache_ref_t {
+     int fd;
+@@ -65,5 +68,6 @@ void h2o_filecache_close_file(h2o_filecache_ref_t *ref);
+ struct tm *h2o_filecache_get_last_modified(h2o_filecache_ref_t *ref, char *outbuf);
+ size_t h2o_filecache_get_etag(h2o_filecache_ref_t *ref, char *outbuf);
+ int h2o_filecache_compare_etag_strong(const char *tag1, size_t tag1_len, const char *tag2, size_t tag2_len);
++int h2o_filecache_in_nix_store(h2o_filecache_ref_t *ref);
+ 
+ #endif
+diff --git a/lib/common/filecache.c b/lib/common/filecache.c
+index 2ba4679d..452853ca 100644
+--- a/lib/common/filecache.c
++++ b/lib/common/filecache.c
+@@ -24,9 +24,13 @@
+ #include <fcntl.h>
+ #include <stddef.h>
+ #include <unistd.h>
++#include <limits.h>
++#include <stdlib.h>
++#include <openssl/sha.h>
+ #include "khash.h"
+ #include "h2o/memory.h"
+ #include "h2o/filecache.h"
++#include "h2o/string_.h"
+ 
+ KHASH_SET_INIT_STR(opencache_set)
+ 
+@@ -163,9 +167,48 @@ struct tm *h2o_filecache_get_last_modified(h2o_filecache_ref_t *ref, char *outbu
+ size_t h2o_filecache_get_etag(h2o_filecache_ref_t *ref, char *outbuf)
+ {
+     assert(ref->fd != -1);
++
++    /* Nix store files have mtime = 1 */
++    if (ref->_etag.len == 0 && ref->st.st_mtime == 1) {
++        static const char nix_store_prefix[] = "@storeDir@/";
++        static const size_t nix_store_prefix_len = sizeof(nix_store_prefix) - 1;
++
++        /* Even if the path doesn't start with the Nix store path,
++         * mtime = 1 means it's most likely a link into the store, e.g.:
++         *
++         *     /var/www/riot.example.com
++         *       -> /nix/store/00000000000000000000000000000000-riot-web
++         *
++         * We check ref->_path for the prefix before calling realpath to
++         * avoid the overhead of the multiple syscalls involved (yes,
++         * it's measurable!) */
++        char *respath = NULL;
++        if (strncmp(ref->_path, nix_store_prefix, nix_store_prefix_len) == 0 ||
++            ((respath = realpath(respath, NULL)) != NULL &&
```

Comment:
Oops. Yes, you're right; I guess the test is working because it has a direct `/nix/store` prefix rather than going through a symlink, so it benefits from the happy-path optimization. I'll correct the patch and extend the test to check this case tomorrow.